consensus/executor/executor.go

```diff
@@ -0,0 +1,220 @@
+/*
+Copyright IBM Corp. 2016 All Rights Reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+		 http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package executor
+
+import (
+	"fmt"
+
+	"github.com/hyperledger/fabric/consensus"
+	"github.com/hyperledger/fabric/consensus/obcpbft/events"
+	"github.com/hyperledger/fabric/core/peer/statetransfer"
+	pb "github.com/hyperledger/fabric/protos"
+
+	"github.com/op/go-logging"
+)
+
+var logger *logging.Logger // package-level logger
+
+func init() {
+	logger = logging.MustGetLogger("consensus/executor")
+}
+
+// PartialStack contains the ledger features required by the executor.Coordinator
+type PartialStack interface {
+	consensus.LegacyExecutor
+	GetBlockchainInfo() *pb.BlockchainInfo
+}
+
+type coordinatorImpl struct {
+	manager         events.Manager              // Maintains event thread and sends events to the coordinator
+	rawExecutor     PartialStack                // Does the real interaction with the ledger
+	consumer        consensus.ExecutionConsumer // The consumer of this coordinator which receives the callbacks
+	stc             statetransfer.Coordinator   // State transfer instance
+	batchInProgress bool                        // Are we mid execution batch
+	skipInProgress  bool                        // Are we mid state transfer
+}
+
+// NewCoordinatorImpl creates a new executor.Coordinator
+func NewImpl(consumer consensus.ExecutionConsumer, rawExecutor PartialStack, stps statetransfer.PartialStack) consensus.Executor {
+	co := &coordinatorImpl{
+		rawExecutor: rawExecutor,
+		consumer:    consumer,
+		stc:         statetransfer.NewCoordinatorImpl(stps),
+		manager:     events.NewManagerImpl(),
+	}
+	co.manager.SetReceiver(co)
+	return co
+}
+
+// ProcessEvent is the main event loop for the executor.Coordinator
+func (co *coordinatorImpl) ProcessEvent(event events.Event) events.Event {
+	switch et := event.(type) {
+	case executeEvent:
+		if logger.IsEnabledFor(logging.DEBUG) {
+			if logger.IsEnabledFor(logging.DEBUG) {
+				logger.Debug("Executor is processing an executeEvent")
+			}
+		}
+		if co.skipInProgress {
+			logger.Error("FATAL programming error, attempted to execute a transaction during state transfer")
+			return nil
+		}
+
+		if !co.batchInProgress {
+			if logger.IsEnabledFor(logging.DEBUG) {
```

Comment:
Yes, on string literals, there is no garbage or performance bottleneck, I threw these on at the last minute because of the comments from #1749 and should have thought about them a little more.

I've pushed a commit fixing these things.